Init flux plugin

ashu8912 opened the pull request.

ashu8912 commented:
<img width="1707" alt="Screenshot 2024-08-20 at 1 32 53 PM" src="https://github.com/user-attachments/assets/52c79b06-fae3-423c-bb91-b503ce0e0e1b">
<img width="1710" alt="Screenshot 2024-08-20 at 1 32 14 PM" src="https://github.com/user-attachments/assets/91c4cf5b-78e4-41be-857a-908d5b40febb">
<img width="1710" alt="Screenshot 2024-08-20 at 1 31 57 PM" src="https://github.com/user-attachments/assets/b649d09d-f0d8-4cae-8548-b3fa1e6ee736">
<img width="1709" alt="Screenshot 2024-08-20 at 1 31 42 PM" src="https://github.com/user-attachments/assets/b1a10379-2ec2-4c89-b360-fe64738d4b29">
<img width="1707" alt="Screenshot 2024-08-20 at 1 31 33 PM" src="https://github.com/user-attachments/assets/11c19665-5cdf-4a5c-a01b-2cc9a62adf3e">
<img width="1708" alt="Screenshot 2024-08-20 at 1 31 22 PM" src="https://github.com/user-attachments/assets/f8ede4a4-a793-42f3-b168-43ca410db9c7">
<img width="1705" alt="Screenshot 2024-08-20 at 1 31 04 PM" src="https://github.com/user-attachments/assets/234fc20c-5ad2-443f-9919-be579b142771">
<img width="1710" alt="Screenshot 2024-08-20 at 1 29 20 PM" src="https://github.com/user-attachments/assets/dc5c007f-0c8c-45e7-ac88-497dbbb8f85e">
<img width="1709" alt="Screenshot 2024-08-20 at 1 29 05 PM" src="https://github.com/user-attachments/assets/f4639c55-56e9-455c-8caf-56214b180d44">

ashu8912 commented:
fixes #74

mgalesloot commented:
A comparison between HeadLamp flux plugin and Weave Gitops is provided below. The focus is on Kustomization and HelmRelease. 

| Page |  Review  |
| ------------- | ------------- |
| applications  | Weave GitOps shows helmreleases and kustomizations in a single table. |
| applications  | Weave shows kustomizations and helmreleases with errors on top which is nice for monitoring. The flip side is that the entries are continuously being sorted which is annoying. |  
| applications  | Headlamp: width of the colums do not resize to content. Other headlamp pages (e.g list of pods) have better column widths.  |
| applications  | Headlamp: Icon Ready/Failed is quite large making the table untidy.  I would prefer the same styling as in other HeadLamp pages e.g. Workloads/Pods/Status | 
| applications  | Weave shows `Tenant` and `Verified` fields which are not shown in Headlamp.  |
| applications  | Headlamp shows `Revision` and `Last Updated`, which are not shown in Weave. Helmrelease does not have a `lastAppliedRevision`,  maybe we should use `lastAttemptedRevision` for helmrelease |   
| applications  | Weave shows checkboxes supported with actions for Sync, Suspend, Resume.   |
| applications  | Weave does not provide pagination. |
| kustomization | Weave shows more information in a collapsable panel, however quite a few fields are not shown (creation, prune, force, annotations, ...) | 
| kustomization | Weave shows icon and status info in the header, e.g. "GitRepository.source.toolkit.fluxcd.io "does-not-exist" not found" | 
| kustomization | Weave shows `Last Updated 1 minute ago` in the top right of the page | 
| kustomization | Weave shows icon and health checks in the header (but I doubt you can trust it to be correct) |
| kustomization | SourceRef is in HeadLamp shown as JSON, in Weave the ref to source is shown as a hyperlink |  
| kustomization/Inventory | Weave Gitops calls the Inventory "Details" |
| kustomization/Inventory | Order of fields is name/kind/namespace (headlamp has name/namespace/kind) |
| kustomization/Inventory | Weave shows a message field (e.g. "ReplicaSet "source-controller-54c89dcbf6" has successfully progressed.")
| kustomization/Inventory | Weave shows the actual container images |  
| kustomization/Inventory | HeadLamp does not show Custom resources such as cert-manager.io/ClusterIssuer | 
| kustomization/Inventory | Weave shows a Healthy column (not always useful, my test shows a healthy icon with the message "Deployment does not have minimum availability." ) | 
| kustomization/Events | Default sort order in Weave is `Last event first`.  |
| kustomization/Conditions | Weave does not show conditions | 
| kustomization/Graph | Weaves graph looks nice but impossible to use on smaller screens.  |  
| kustomization/Dependencies | Weave shows dependencies as a graph. Dependencies are red if not ready. Headlamp shows dependencies, but there is no navigation link and the status is not shown.  |   
| kustomization/Yaml | Weave provides read only YAML, HeadLamp direct;ly goes into edit mode. | 
| kustomization/Violations | Weave shows violations from the Weave Enterprise Policy Agent. | 
| helmrelease | Most findings for kustomization apply to helmrelease as well. | 
| helmrelease | Weave shows values only in the YAML. | 
| helmrelease | Weave shows `Last attempted revision` , `Last applied revision` | 
| helmrelease | Weave does not show `Reconcile strategy` (dont understand why) | 
| helmrelease/Inventory | Nothing is shown in Weave nor Headlamp. Please note `Flux tree` shows details for helmreleases. |

stefanprodan commented:
I've been reading the code and for HelmRelease it assumes that the chart is always embedded in the HelmRelease while in the Flux API this is just one of the possibilities. For charts hosted in OCI repositories this is how it looks:

```yaml
apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: OCIRepository
metadata:
  name: podinfo
  namespace: default
spec:
  interval: 10m
  url: oci://ghcr.io/stefanprodan/charts/podinfo
  ref:
    semver: ">6.0.0"
---
apiVersion: helm.toolkit.fluxcd.io/v2
kind: HelmRelease
metadata:
  name: podinfo
  namespace: default
spec:
  interval: 10m
  releaseName: podinfo
  chartRef:
    kind: OCIRepository
    name: podinfo
```

Note how there is no `.spec.chart.spec` but instead `.spec.chartRef`.

ashu8912 commented:
> I've been reading the code and for HelmRelease it assumes that the chart is always embedded in the `.spec.chart` while in the Flux API this is just one of the possibilities. For charts hosted in OCI repositories this is how it looks:
> 
> ```yaml
> apiVersion: source.toolkit.fluxcd.io/v1beta2
> kind: OCIRepository
> metadata:
>   name: podinfo
>   namespace: default
> spec:
>   interval: 10m
>   url: oci://ghcr.io/stefanprodan/charts/podinfo
>   ref:
>     semver: ">6.0.0"
> ---
> apiVersion: helm.toolkit.fluxcd.io/v2
> kind: HelmRelease
> metadata:
>   name: podinfo
>   namespace: default
> spec:
>   interval: 10m
>   releaseName: podinfo
>   chartRef:
>     kind: OCIRepository
>     name: podinfo
> ```
> 
> Note how there is no `.spec.chart.spec` but instead `.spec.chartRef`.

Thanks @stefanprodan for pointing that out, i will add a patch for this

ashu8912 commented:
Hii @stefanprodan I have added a fix for the above in this same PR

ashu8912 commented:
I have also taken some of @mgalesloot comments and added fix for them as well

stefanprodan commented on flux-plugin/src/flux-applications/application.tsx:
```diff
@@ -0,0 +1,390 @@
+    in the format '<namespace>_<name>_<group>_<kind>'.
+    */
+    const parsedID = id.split('_');
+    const namespace = parsedID[0];
```
Not all resources have a namespace, I suggest testing this on cluster where you've run `flux bootstrap` and see if the `flux-system` Kustomization renders here the ClusterRoles, Namespaces, etc.

ashu8912 replied:
The K8s API we use handles this, but i was missing passing an undefined namespace which was an issue, thanks @stefanprodan for pointing this out.

Aubermean commented:
How can I install this? Could you release it via Artifact Hub or somewhere else so it can be installed in the Desktop App by enabling unofficial repositories? I tried to download the fork and build the plugin but I am having issues installing ` @kinvolk/headlamp-plugin`...

Aubermean commented:
It seems without building the plugins and building the desktop application, there isn't a way to install the Flux plugin. Can we get this PR merged so others can start using it? In any case, the plugin isn't enabled by default so a first release shouldn't cause breaking changes right? Thanks! @stefanprodan @ashu8912

Aubermean commented:
I have managed to get the plugin installed. I am not a node or javascript guy so it took some strong wifi and a lot of dependencies. It's awesome, thank you, great work guys.

I noticed some things, and had some ideas. But instead of writing them here, I created a new issue for each.

I have never once used a Flux UI like Weaveworks, Flamingo, Capacitor etc. so maybe someone who has (@mgalesloot ?) would like to chime in.

ashu8912 commented:
Thank you @Aubermean for reporting these issues, we will take a look at each one of them.

Aubermean commented:
Thanks @ashu8912, I wish I could do more to contribute but I am just a self-employed guy learning GitOps thanks to @stefanprodan's awesome D1 Architecture PDF/repos :)

Aubermean commented:
For my own reference and for others, this is how to quickly pull, build, and install the flux plugin (or any plugin really) on macOS:

```bash
mkdir -p ~/Downloads/headlamp-plugins
git clone https://github.com/headlamp-k8s/plugins.git ~/Downloads/headlamp-plugins
cd ~/Downloads/headlamp-plugins
git checkout add-flux-plugin
cd flux-plugin
asdf install nodejs 22.8.0
asdf local nodejs 22.8.0
npm install
npm install -g @kinvolk/headlamp-plugin
npx @kinvolk/headlamp-plugin build .
rm -rf ~/Library/Application\ Support/Headlamp/plugins/flux
mkdir -p ~/Library/Application\ Support/Headlamp/plugins/flux
cp package.json dist/* ~/Library/Application\ Support/Headlamp/plugins/flux
rm -rf ~/Downloads/headlamp-plugins
```

Smana commented:
I'll give it a try, thank you @Aubermean

Smana commented:
> I'll give it a try, thank you @Aubermean

I got it working with tiny changes in Linux
```console
mkdir -p ~/.config/Headlamp/plugins/flux
cp package.json dist/* ~/.config/Headlamp/plugins/flux
```

![Screenshot from 2024-09-06 21-27-48](https://github.com/user-attachments/assets/75453733-2389-450c-b3fa-ec5511067b89)

stefanprodan commented:
From the last printscreen looks like the HelmRelease revision is broken. @ashu8912 here are the json paths to find the revision:

- HelmRelease 
https://github.com/fluxcd/flux2-monitoring-example/blob/5bf921a7b553df4dde59a186549f450b6c306be8/monitoring/controllers/kube-prometheus-stack/kube-state-metrics-config.yaml#L72

- Kustomization
https://github.com/fluxcd/flux2-monitoring-example/blob/5bf921a7b553df4dde59a186549f450b6c306be8/monitoring/controllers/kube-prometheus-stack/kube-state-metrics-config.yaml#L53

stefanprodan commented:
To display the HelmRelease inventory you need to query the cluster for all objects with the following labels:

```
helm.toolkit.fluxcd.io/name: <HELM RELEASE NAME>
helm.toolkit.fluxcd.io/namespace: <HELM RELEASE NAMESPACE>
```

Aubermean commented:
<img width="1393" alt="image" src="https://github.com/user-attachments/assets/a698a292-e475-42fd-adbc-03ae3a8c9138">

Also showing Ready status when its not Ready

stefanprodan commented:
> Also showing Ready status when its not Ready

The `Ready` condition has 3 possible statuses:
- `status: Unknown` - we should display `Reconciling`
- `status: True` - we should display `Ready`
- `status: Flase` - we should display `Failing`

ashu8912 commented:
> <img alt="image" width="1393" src="https://private-user-images.githubusercontent.com/58298118/365365323-a698a292-e475-42fd-adbc-03ae3a8c9138.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU5NDMyNDksIm5iZiI6MTcyNTk0Mjk0OSwicGF0aCI6Ii81ODI5ODExOC8zNjUzNjUzMjMtYTY5OGEyOTItZTQ3NS00MmZkLWFkYmMtMDNhZTNhOGM5MTM4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEwVDA0MzU0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI5M2Q1YzFmNzM4NjQzZmM4NGQyNzU1MGQ0ZTk5ZWRjNDM3YzA1MjljNTNkNTVkYjNiYmZjMGZiZWZlMDU1ZWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.snBnbz4P7XoNonlTwxzoxliahKfo2-_jIGOixLB0n-g">
> Also showing Ready status when its not Ready

Fixed this now @Aubermean

ashu8912 commented:
> To display the HelmRelease inventory you need to query the cluster for all objects with the following labels:
> 
> ```
> helm.toolkit.fluxcd.io/name: <HELM RELEASE NAME>
> helm.toolkit.fluxcd.io/namespace: <HELM RELEASE NAMESPACE>
> ```

This is now fixed we show All the inventory items now, The CustomResources were missing but i have added them as well

ashu8912 commented:
> > Also showing Ready status when its not Ready
> 
> The `Ready` condition has 3 possible statuses:
> 
> * `status: Unknown` - we should display `Reconciling`
> * `status: True` - we should display `Ready`
> * `status: False` - we should display `Failing`

Done, fixed this thanks @stefanprodan

ashu8912 commented:
> From the last printscreen looks like the HelmRelease revision is broken. @ashu8912 here are the json paths to find the revision:
> 
> * HelmRelease
>   https://github.com/fluxcd/flux2-monitoring-example/blob/5bf921a7b553df4dde59a186549f450b6c306be8/monitoring/controllers/kube-prometheus-stack/kube-state-metrics-config.yaml#L72
> * Kustomization
>   https://github.com/fluxcd/flux2-monitoring-example/blob/5bf921a7b553df4dde59a186549f450b6c306be8/monitoring/controllers/kube-prometheus-stack/kube-state-metrics-config.yaml#L53
> 
> Note that in both cases, the json path may be missing if the Flux resource failed to reconcile at creation time.

I actually didn't get this so for the revision what property can i actually rely on?? In the above case i used `status.lastAppliedVersion`

stefanprodan commented:
@ashu8912 for both HelmRelease and Kustomization there is a `status.lastAttemptedRevision` that you can use.

ashu8912 commented:
> @ashu8912 for both HelmRelease and Kustomization there is a `status.lastAttemptedRevision` that you can use. HelmRelease does not have `status.lastAppliedVersion` and that's why it never displays.

done thanks @stefanprodan

kingdonb commented:
I have rebuilt the `kingdonb/plugins:canary` image that I was using to test, I will spend some more time with the Headlamp UI this week 👍 it's great to see more people trying it out! Hopefully we can get this polished enough to promote into the catalog soon, so it can be easier for end-users to install it.

It looks like it is still a bit of a lift to install it either locally or in the HelmRelease.

Anyway, for those that might wish to try the Helm in-cluster version of Headlamp with the Flux plugin, my WIP first attempt at building it for easy install is here - the Dockerfile needed a few extra things, namely `npm install` to be run in each plugin

(I excluded the other plugins in the `plugins` repo to make it easier, I don't think that's a good change to merge upstream):

* https://github.com/kingdonb/plugins/tree/build-plugins-docker

You need to use an `initContainers` config to load the plugins image into an existing Headlamp helm installation, I won't publish the config I used for now because it isn't finished, and it's already documented how to do this here:

* https://headlamp.dev/blog/2022/10/20/best-practices-for-deploying-headlamp-with-plugins#deploying-headlamp-with-plugins

If anyone is struggling to get this deployed, you're welcome to reach me here or on the CNCF slack. (I'll be happy to share!)

mgalesloot commented:
Deploying in-cluster does not require to build your own image. I have it running with the following helm values.
```
 config:
  pluginsDir: /build/plugins
initContainers:
  - command:
      - /bin/sh
      - -c
      - mkdir -p /build/plugins/flux-plugin
        && wget https://github.com/mgalesloot/headlamp-flux-plugin-release/releases/download/latest/main.js -O /build/plugins/flux-plugin/main.js
        && wget https://github.com/mgalesloot/headlamp-flux-plugin-release/releases/download/latest/package.json -O /build/plugins/flux-plugin/package.json
    image: busybox
    name: headlamp-plugins
    volumeMounts:
      - mountPath: /build/plugins
        name: headlamp-plugins
volumeMounts:
  - mountPath: /build/plugins
    name: headlamp-plugins
volumes:
  - name: headlamp-plugins
    emptyDir: {}
```
This release on github is created manually, after pulling the code and run `npm run build`. Obviously this could be done better with a github action. I would recommend to make a github action in headlamp-plugin repo to build a latest  for easy testing.

kingdonb commented:
Using an image will allow me to instrument the deployment with the Flux Image Update Automation, and gives us an audit trail for rollback in case something breaks on an intermediate deployment, so I will probably do it that way, but this is neat! Thanks for sharing!

kingdonb commented:
I got the image running on my cluster, the cluster had some issues I had to resolve. They're fixed!

I can see some things that were not working before, are working now. The issue that I reported here (with HelmReleases) is fixed:

https://github.com/headlamp-k8s/headlamp/issues/1872#issuecomment-2304864932

These are the HelmRelease values that I used, so I can incorporate whatever plugins are in the image that I built:

```yaml
config:
  pluginsDir: /build/plugins
initContainers:
  - command:
      - /bin/sh
      - -c
      - mkdir -p /build/plugins && cp -r /plugins/* /build/plugins/
    image: kingdonb/plugins:canary
    imagePullPolicy: Always
    name: headlamp-plugins
    volumeMounts:
      - mountPath: /build/plugins
        name: headlamp-plugins
persistentVolumeClaim:
  accessModes:
    - ReadWriteMany
  enabled: true
  size: 1Gi
volumeMounts:
  - mountPath: /build/plugins
    name: headlamp-plugins
volumes:
  - name: headlamp-plugins
    persistentVolumeClaim:
      claimName: headlamp
```

I copied and pasted these values directly out of the HelmRelease UI in Headlamp 😝 

There is no OIDC config in this values.yaml, it's just using the UI-provided instructions about how to create a token for a service account.

I don't want to give 👍  to release this in case other people need to review it first, but my next steps are to configure ingress, oidc, and stealth-publish my own full `HelmRelease`/`HelmRepository` yamls so Flux users can deploy this with GitOps and test it out. And it looks like people are already finding it on their own; I think it's moving in the right direction!

mgalesloot commented:
It seems you are on the same track as I am. Flux + KeyCloak + HeadLamp. In case you are interested, here is my configuration: https://gitlab.com/matthijs.galesloot/k8s-dd/-/tree/main/components/headlamp. If you need help let me know. I have this stuff running on a test site https://headlamp.kubebeam.nl, you can sign in if you like with demo/demo.

samos667 commented:
I get this error when trying to display a kustomization:
![image](https://github.com/user-attachments/assets/16dbac01-2de5-4666-9faa-382ed39fb967)

ashu8912 commented:
Thanks @samos667 for reporting this we are looking into this and will add a fix.

ashu8912 commented:
Hii @samos667  I have pushed a small change which should fix this can you confirm??

samos667 commented:
Wow that was quick, t4t !
I will test it ASAP as I get home

samos667 commented:
Is like only kustomation who don't declare helmrelease is working
![image](https://github.com/user-attachments/assets/6eacb66a-a248-42d6-b3fa-f899f704cbd2)

But for all others who declare helmelease, it's crashing again
![image](https://github.com/user-attachments/assets/24e31900-c752-4f9c-b2d1-800317675ada)

Smana commented:
@kingdonb would that be interesting if we mount the plugins with [Volumes based on OCI artifacts](https://kubernetes.io/blog/2024/08/16/kubernetes-1-31-image-volume-source/) ?

Smana commented:
I used your method @mgalesloot and that works fine thank you!
see [here](https://github.com/Smana/cloud-native-ref/tree/feat_flux_headlamp/tooling/base/headlamp). The only change is that I use Gateway API instead.

![Screenshot from 2024-09-13 12-29-55](https://github.com/user-attachments/assets/8743fb63-f78e-405f-8f59-9c0bc6f37e48)

I'm just wondering if there's another way to add plugins. Btw that's pretty hard to locate all the available headlamp plugins, or I don't have the proper link.

Smana commented:
@mgalesloot oh thanks I didn't noticed there was a type `headlamp plugins`. I'll give a try to your plugin!

kingdonb commented:
> @kingdonb would that be interesting if we mount the plugins with [Volumes based on OCI artifacts](https://kubernetes.io/blog/2024/08/16/kubernetes-1-31-image-volume-source/) ?

Ideally we're not building a multi-arch image just to include some javascript files, yes, I think OCI artifacts as volumes would help!

ashu8912 commented:
> Is like only kustomation who don't declare helmrelease is working ![image](https://private-user-images.githubusercontent.com/50653464/367068468-6eacb66a-a248-42d6-b3fa-f899f704cbd2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY3MjYyNDMsIm5iZiI6MTcyNjcyNTk0MywicGF0aCI6Ii81MDY1MzQ2NC8zNjcwNjg0NjgtNmVhY2I2NmEtYTI0OC00MmQ2LWIzZmEtZjg5OWY3MDRjYmQyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE5VDA2MDU0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ5OWQ1NDQ4NDQ1M2U4NTMwOTY2NWZlMzc4YmYzNTdjZDdmNWQyY2UzYTA5NjQwOGRhMGQ3YTdmZDFkODY2YjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.AwS2O73BNWduiRA0P8kUVlP0HK3fIBXTiqkeuyrh0JI)
> 
> But for all others who declare helmelease, it's crashing again ![image](https://private-user-images.githubusercontent.com/50653464/367068775-24e31900-c752-4f9c-b2d1-800317675ada.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY3MjYyNDMsIm5iZiI6MTcyNjcyNTk0MywicGF0aCI6Ii81MDY1MzQ2NC8zNjcwNjg3NzUtMjRlMzE5MDAtYzc1Mi00ZjljLWIyZDEtODAwMzE3Njc1YWRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE5VDA2MDU0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ1OGVmYjRhZGUwZWE1Y2U4NThiMjJhNWIyOTZmZmM3OTM3MzdlYzg1YTQxN2Y2YWVmNjc0MDdmMjA4N2MwOTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.EeGNSS28xYzMmcqVEVni76smF8huVEqLjgZRAZMVjKU)

Hii @samos667 is the HelmRelease thing fixed for you now, i have added a couple of new changes, also i am not able to reproduce this, can you give me a example kustomization where this fails. Thank you

mgalesloot commented:
Hi I have made a tar build in a fork of the project for testing purposes: https://github.com/Kubebeam/headlamp-plugins/releases/download/latest/flux-plugin-latest.tar.gz

mgalesloot commented:
The default sort order for Helm release and Kustomizations is currently by Name. The disadvantage is that failing Kustomizations might only be shown on page 2 of the table. Rather the default sort order should be on status, with Failed on top, similar as we have had in Weave Gitops.

mgalesloot commented:
Some more findings: 
- The sort order in Kustomizations and HelmReleases should show Not Ready and Suspended on top, supporting quick monitoring
- If a kustomization/helmrelease is suspended, the status field should show this (currently it indicates "Ready") 
- Custom resources do not show up in the inventory (example CertManager ClusterIssuer) 
- Sort order of events should be last event on top
- Revision and Message in the Kustomization table show the same info. Suggestion: hide the revision at this spot just like Weave GitOps does. 
- Tables for Kustomizations and HelmReleases would benefit from column width indicators. it would allow the message to take more space,. 
- Title of pages should be Capitalized and in a phrase (Image Automations and not image-automations)

Smana commented:
I just noticed that, when a kustomization is being reconciled, I get the status "Failed"
![image](https://github.com/user-attachments/assets/37d4eda1-c514-466c-8f19-dc4d38d0c95d)

ashu8912 commented:
> I just noticed that, when a kustomization is waiting for a dependency, I get the status "Failed" ![image](https://private-user-images.githubusercontent.com/4101366/370179986-37d4eda1-c514-466c-8f19-dc4d38d0c95d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxNjc2ODAsIm5iZiI6MTcyNzE2NzM4MCwicGF0aCI6Ii80MTAxMzY2LzM3MDE3OTk4Ni0zN2Q0ZWRhMS1jNTE0LTQ2NmMtOGYxOS1kYzRkMzhkMGM5NWQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjRUMDg0MzAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTM1MGEwYTIyYWMyM2MwNTFlNDIwZWEzY2Q2MzhiOGU5YjA1ZWM1ZTAwNjc3ZjkzODQ2NmY4NmY0NzllMzBkOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.yfcsen3Chvs2wXjkkwMaLmBSD2lf5WwlC4scwhNQDXk)

Hii @Smana it follows this https://github.com/headlamp-k8s/plugins/pull/75#issuecomment-2335270796

ashu8912 commented:
> Some more findings:
> 
> * The sort order in Kustomizations and HelmReleases should show Not Ready and Suspended on top, supporting quick monitoring
> * If a kustomization/helmrelease is suspended, the status field should show this (currently it indicates "Ready")
> * Custom resources do not show up in the inventory (example CertManager ClusterIssuer)
> * Sort order of events should be last event on top
> * Revision and Message in the Kustomization table show the same info. Suggestion: hide the revision at this spot just like Weave GitOps does.
> * Tables for Kustomizations and HelmReleases would benefit from column width indicators. it would allow the message to take more space,.
> * Title of pages should be Capitalized and in a phrase (Image Automations and not image-automations)

Hii @mgalesloot thanks for this list
1. Customization the sort order where it follows a certain order i will check if it can be done through the Table Component we have otherwise we will have to think of a new solution for this.
2. Yeah The status follows https://github.com/headlamp-k8s/plugins/pull/75#issuecomment-2335270796 but if we want custom messages like reconciling or suspended we can add that too.
3. Umm i added a patch a couple of days back for this, we have to seperately handle CustomResources, I thought this was fixed, i will take a look into this again.
4. Got it
5. Got it
6. If a message is big enough we now show a expand icon to view the full message
7. Got it

joaquimrocha commented on flux-plugin/src/flux-image-automation/image-automation.tsx:
```diff
@@ -0,0 +1,210 @@
+  // The last segment is the name
+  const name = segments[segments.length - 1];
+
+  function getType() {
+    if (type === 'imagerepositories') {
+      return IMAGE_REPOSITORIES;
+    }
+    if (type === 'imagepolicies') {
+      return IMAGE_POLICIES;
+    }
+    return IMAGE_UPDATE_AUTOMATIONS;
+  }
+
+  function getKind() {
+    if (type === 'imagerepositories') {
+      return 'ImageRepository';
+    }
+    if (type === 'imagepolicies') {
+      return 'ImagePolicy';
+    }
+    return 'ImageUpdateAutomation';
+  }
```
The `IMAGE_REPOSITORIES` and similar vars is only used once, AFAICS. And you do a similar logic for the kind.
So why not have an object with them:
```typescript
const fluxImageInfo = {
  imagerepositories: {
    type: 'imagerepositories.image.toolkit.fluxcd.io',
    kind: 'ImageRepository',
  },
  ...
}
```

I think this becomes easier to maintain.

joaquimrocha commented on flux-plugin/src/flux-image-automation/image-automation.tsx:
```diff
@@ -0,0 +1,210 @@
+            {
+              header: 'Message',
+              accessorFn: item => (
+                <ShowHideLabel labelId={item?.metadata?.uid || ''}>
```
Use `??` when it's checking for undefined.

joaquimrocha commented on flux-plugin/src/flux-image-automation/image-automations.tsx:
```diff
@@ -0,0 +1,270 @@
+
+export default function ImageAutomation() {
+  const CRD = K8s.ResourceClasses.CustomResourceDefinition;
+  CRD.apiEndpoint = apiFactory(
+    ['apiextensions.k8s.io', 'v1', 'customresourcedefinitions'],
+    ['apiextensions.k8s.io', 'v1beta1', 'customresourcedefinitions'],
+    ['apiextensions.k8s.io', 'v1beta2', 'customresourcedefinitions']
+  );
```
Would it work to simply add the 3rd value if there's no v1beta2 in the apiEndpoint (this should be checked only once)? I am mentioning this because we will add this new API version and maybe more in the future, and then we have to update the plugin for it to work with versions past these 3.

ashu8912 replied:
updated

joaquimrocha commented on flux-plugin/src/flux-runtime/runtime.tsx:
```diff
@@ -0,0 +1,109 @@
+import { Link as MuiLink } from '@mui/material';
+
+export default function FluxRunTime(props) {
+  const [pods, error] = K8s.ResourceClasses.Pod.useList();
+  const [crds] = K8s.ResourceClasses.CustomResourceDefinition.useList();
+  const helmController = pods?.filter(pod => pod.metadata.labels['app'] === 'helm-controller');
+  const kustomizeController = pods?.filter(
+    pod => pod.metadata.labels['app'] === 'kustomize-controller'
+  );
+  const notificationController = pods?.filter(
+    pod => pod.metadata.labels['app'] === 'notification-controller'
+  );
+  const sourceController = pods?.filter(pod => pod.metadata.labels['app'] === 'source-controller');
+  const imageReflectorController = pods?.filter(
+    pod => pod.metadata.labels['app'] === 'image-reflector-controller'
+  );
+  const imageAutomationController = pods?.filter(
+    pod => pod.metadata.labels['app'] === 'image-automation-controller'
+  );
```
We could add this filtering to the list query. But maybe now it doesn't matter so much, with react-query.

joaquimrocha commented on flux-plugin/src/flux-sources/source.tsx:
```diff
@@ -0,0 +1,199 @@
+function CustomResourceDetailView(props) {
+  const { name, namespace, resource } = props;
+  const [cr, setCr] = React.useState(null);
+  //const [openDialog, setOpenDialog] = React.useState(false);
```
Please delete comments like these.

joaquimrocha commented on flux-plugin/src/flux-sources/source.tsx:
```diff
@@ -0,0 +1,199 @@
+
+function Events(props) {
+  const { namespace, name, cr } = props;
+  const [events] = Event?.default.useList({
+    namespace: namespace,
+    fieldSelector: `involvedObject.name=${name},involvedObject.kind=${cr?.jsonData.kind}`,
+  });
```
We have `Event.objectEvents` for this.

joaquimrocha commented on flux-plugin/src/index.tsx:
```diff
@@ -0,0 +1,136 @@
+import { FluxImageAutomationDetailView } from './flux-image-automation/image-automation';
+import Notification from './flux-notifications/notification';
+
+// Below are some imports you may want to use.
+//   See README.md for links to plugin development documentation.
+// import { SectionBox } from '@kinvolk/headlamp-plugin/lib/CommonComponents';
+// import { K8s } from '@kinvolk/headlamp-plugin/lib/K8s';
+// import { Typography } from '@mui/material';
+
```
Please delete this.

joaquimrocha commented on flux-plugin/src/resoureceMap/ClusterResources.tsx:
```diff
@@ -0,0 +1,128 @@
+import Deployment from '@kinvolk/headlamp-plugin/lib/k8s/deployment';
+import Ingress from '@kinvolk/headlamp-plugin/lib/k8s/ingress';
+import Job from '@kinvolk/headlamp-plugin/lib/k8s/job';
+//import MutatingWebhookConfiguration from '@kinvolk/headlamp-plugin/lib/k8s/mutatingWebhookConfiguration';
```
Please delete comments like this.

joaquimrocha commented on flux-plugin/src/resoureceMap/EventsDetails.tsx:
```diff
@@ -0,0 +1,35 @@
+      <Typography variant="h6">Events</Typography>
+
+      <SimpleTable
+        data={events?.filter(it => it.involvedObject.uid === resource.metadata.uid) ?? []}
```
Maybe you want to keep a memo of this, or use the query field in useList.

joaquimrocha commented on flux-plugin/src/resoureceMap/GraphLayout.tsx:
```diff
@@ -0,0 +1,206 @@
+      .then(node => {
+        return node;
+      })
+      // .then(node =>
+      //   // another layout pass to make sure nodes are not overlapping
+      //   elk.layout(node, { layoutOptions: { 'elk.algorithm': 'org.eclipse.elk.sporeOverlap' } })
+      // )
```
Delete comments.

ashu8912 replied:
done

joaquimrocha commented on flux-plugin/README.md:
```diff
@@ -0,0 +1,15 @@
+# flux-plugin
+
+This is the default template README for [Headlamp Plugins](https://github.com/headlamp-k8s/headlamp).
+
+- The description of your plugin should go here.
+- You should also edit the package.json file meta data (like name and description).
+
+## Developing Headlamp plugins
+
+For more information on developing Headlamp plugins, please refer to:
+
+- [Getting Started](https://headlamp.dev/docs/latest/development/plugins/), How to create a new Headlamp plugin.
+- [API Reference](https://headlamp.dev/docs/latest/development/api/), API documentation for what you can do
+- [UI Component Storybook](https://headlamp.dev/docs/latest/development/frontend/#storybook), pre-existing components you can use when creating your plugin.
+- [Plugin Examples](https://github.com/headlamp-k8s/headlamp/tree/main/plugins/examples), Example plugins you can look at to see how it's done.
```
Let's replace this with a description and a link to instructions on installing Flux.

ashu8912 replied:
done

joaquimrocha reviewed (changes requested):
Rough review but found a few things that could be improved.

kingdonb commented:
I just made some mention of this plugin in headlamp-k8s/headlamp#2342

The plugin does not function when installed in a single-namespace environment, because there is no way to detect if Flux is installed. Could we add a [plugin settings](https://headlamp.dev/blog/2024/05/27/plugin-settings#adding-settings-to-a-plugins) option to skip checking, and assume that Flux is installed?

It's better if there is no option needed, maybe if the plugin detects it does not have permission to list CRDs (I assume how it's checking for Flux installed or not) then it would fall back to assuming Flux is installed? My service account token has the admin role in the namespace, so it can definitely read all the Flux resources in the namespace:

```
% kubectl auth can-i list kustomizations.kustomize.toolkit.fluxcd.io --as system:serviceaccount:test-yebyen:flux
yes
```

Users who work in a single-namespace tenant environment will already have to find the Settings page to configure allowed namespace / default namespace, so I don't think expecting them to find the checkbox in plugin settings is too much to ask.

ashu8912 commented on flux-plugin/src/resoureceMap/ResourceMap.tsx:
removed

kingdonb commented:
I built from the latest push today, I got this message:

<img width="1063" alt="Screenshot 2024-09-30 at 3 29 49 PM" src="https://github.com/user-attachments/assets/7e848594-4c3d-4d1d-8431-dc1f82df07d6">

This is the actual state of the cluster:

<img width="1329" alt="Screenshot 2024-09-30 at 3 30 31 PM" src="https://github.com/user-attachments/assets/7edc65d9-7bd3-4edd-9d61-c8c5b2a08701">

It's not a standard `flux bootstrap` install, this is a Flux 2.3.0 installed from the `flux-operator`

I'm having a look through the code right now to see if I can understand why it doesn't pick up my flux installation, in the `cozy-fluxcd` namespace:

<img width="1061" alt="Screenshot 2024-09-30 at 3 31 44 PM" src="https://github.com/user-attachments/assets/ae1f8648-0ba0-4377-b71c-ac706aaf55d7">

gecube commented:
Hello! 

I tried to install the plugin, but unfortunately it does not work properly.

I used the image from @kingdonb so the final manifest for `headlamp` looked like:

```yaml
---
apiVersion: v1
kind: Namespace
metadata:
  name: headlamp
---
apiVersion: source.toolkit.fluxcd.io/v1beta1
kind: HelmRepository
metadata:
  name: headlamp
  namespace: headlamp
spec:
  interval: 1m0s
  timeout: 1m0s
  url: https://headlamp-k8s.github.io/headlamp/
---
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: headlamp
  namespace: headlamp
spec:
  chart:
    spec:
      chart: headlamp
      sourceRef:
        kind: HelmRepository
        name: headlamp
      version: '*'
  interval: 1m0s
  values:
    ingress:
      enabled: true
      ingressClassName: nginx
      annotations: 
        cert-manager.io/cluster-issuer: letsencrypt-http01-nginx-issuer
      hosts:
      - host: headlamp.my-cool-domain.com
        paths:
        - path: /
          type: ImplementationSpecific
      tls:
      - secretName: headlamp-tls
        hosts:
          - headlamp.my-cool-domain.com
    config:
      pluginsDir: /build/plugins
    initContainers:
    - command:
      - /bin/sh
      - -c
      - mkdir -p /build/plugins && cp -r /plugins/* /build/plugins/
      image: kingdonb/plugins:canary
      imagePullPolicy: Always
      name: headlamp-plugins
      volumeMounts:
      - mountPath: /build/plugins
        name: headlamp-plugins
    persistentVolumeClaim:
      accessModes:
      - ReadWriteOnce
      enabled: true
      size: 1Gi
    volumeMounts:
    - mountPath: /build/plugins
      name: headlamp-plugins
    volumes:
    - name: headlamp-plugins
      persistentVolumeClaim:
        claimName: headlamp
```

I used a different namespace for `headlamp` because of the security considerations, but I think it should be fine.

The image from @kingdonb was also the latest version, so I am sure that the issue is not with the version of plugin.

FluxCD is 2.3.0

The token was obtained like:

```bash
$ kubectl create token headlamp -n headlamp
eyJhbGciOiJSUzI1NiIsImtpZCI6InFnRTAzWDFZWGZfLThzcDl2dzY3UU1YUmx2c011bnVNLUZDZU8xQVZGbGMifQ.eyJhdWQiOlsiaHR0cHM6Ly9rdWJlcm5ldGVzLmRlZmF1bHQuc3ZjLmNsdXN0ZXIubG9jYWwiXSwiZXhwIjoxNzI3NzY1ODQwLCJpYXQiOjE3Mjc3NjIyNDAsImlzcyI6Imh0dHBzOi8va3ViZXJuZXRlcy5kZWZhdWx0LnN2Yy5jbHVzdGVyLmxvY2FsIiwianRpIjoiNGYwZWVlOTctNWJjNi00ZTQ4LThkNmEtMWU4NDlhOWZmZDZlIiwia3ViZXJuZXRlcy5pbyI6eyJuYW1lc3BhY2UiOiJoZWFkbGFtcCIsInNlcnZpY2VhY2NvdW50Ijp7Im5hbWUiOiJoZWFkbGFtcCIsInVpZCI6Ijc1YTlmOWU0LWY2NDUtNGYzNi1iZDFhLTA5NGE0MDY1ZWIzNyJ9fSwibmJmIjoxNzI3NzYyMjQwLCJzdWIiOiJzeXN0ZW06c2VydmljZWFjY291bnQ6aGVhZGxhbXA6aGVhZGxhbXAifQ.Sl2XRmgGoqY1yng2iJTKjYVq3FQ1qVA7hZqwvwMkYqcpiCIqj63okQVkdjSTe0T90iP7z2BTOh6DDhB0SGzXIuzLUJmhPNbZUOEtrAKkzAq9QRf1493GafaNIoyP8C0ud-FhHeVRSahxVp2fNyYYe7P-nD9BxuZiEfEDZdeojQl8dsZ1yiXMMknDLw2TCvP1MZN15fv1M9PtgkQrKeERywe8Tj9gC5ihI18FQ4dW-YDHqoiK2D-iCw6Ao_05Th3KLOVRUp6t9y-9HxsjKqKPQMLFWQ-M0fxjoctgpMVKFkXXAsK89dMDdQyeRUc9P4sHPNtBUr9kSAFc3pJPlnBd1w
```

so there is a discrepancy with the [documentation](https://headlamp.dev/docs/latest/installation/#create-a-service-account-token)

I noticed that the pod metrics are not shown (but we are using VictoriaMetrics stack)

<img width="1357" alt="Screenshot 2024-10-01 at 07 59 16" src="https://github.com/user-attachments/assets/085288fa-d445-41ad-aca0-c5e53305fbc1">

And when clicking on FluxCD on the left bar I am getting:

<img width="1353" alt="Screenshot 2024-10-01 at 08 00 06" src="https://github.com/user-attachments/assets/69a1403d-f39c-4080-a4e6-de169580aea0">

In the same time in browser development tools:

<img width="960" alt="Screenshot 2024-10-01 at 08 00 26" src="https://github.com/user-attachments/assets/c61ef09d-98a2-4e13-a731-1912c07237f8">

<img width="944" alt="Screenshot 2024-10-01 at 08 00 51" src="https://github.com/user-attachments/assets/1ae0cc41-a5a9-4a28-a95c-ca60e787530c">

It is very interesting that the API is working and Flux objects were fetched:

<img width="972" alt="Screenshot 2024-10-01 at 08 01 06" src="https://github.com/user-attachments/assets/0e0324f5-45aa-4858-8185-27176f7de3b6">

I am looking for some advice.

ashu8912 commented:
Thanks for reporting this @gecube we are looking into this, will get back to you soon

stefanprodan commented on flux-plugin/src/flux-runtime/runtime.tsx:
```diff
@@ -0,0 +1,112 @@
+import { Link as MuiLink } from '@mui/material';
+
+export default function FluxRunTime() {
+  const [pods] = K8s.ResourceClasses.Pod.useList();
```
Does this loads all the pods in the cluster? If so this is extremely costly and unnecessary.

I propose we get change this list to show Kubernetes Deployments instead of pods. You can directly filter the Deployments by showing only those with the label `app.kubernetes.io/part-of: flux`.

stefanprodan commented on flux-plugin/src/flux-runtime/runtime.tsx:
```diff
@@ -0,0 +1,112 @@
+export default function FluxRunTime() {
+  const [pods] = K8s.ResourceClasses.Pod.useList();
+  const [crds] = K8s.ResourceClasses.CustomResourceDefinition.useList();
+  const helmController = pods?.filter(pod => pod.metadata.labels['app'] === 'helm-controller');
```
helm-controller is optional, if it doesn't exists this doesn't mean Flux is not installed. The only required controller is source-controller.

stefanprodan commented on flux-plugin/src/flux-runtime/runtime.tsx:
```diff
@@ -0,0 +1,112 @@
+          ]}
+        />
+      </SectionBox>
+      <SectionBox title="Custom Resource Definitions">
```
This table should display the storage versions.

stefanprodan commented:
I suggest we split the Application page into Kustomizations and HelmReleases (dedicated pages). Lots of Flux users don't use Helm at all, and there is no concept of an Application in Flux.

joaquimrocha reviewed (approved):
LGTM! I did squash some new changes:
- Add the status label to the details views (it's a quick way to see the status)
- Fix reconciliation count when the item doesn't change after the count-down
- Hide some tables in KustomizationSingle if the resource is still null

@ashu8912 I think the only thing missing now is the Flux icon being loaded locally instead of from iconify.
